oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+        let generic_handler = HttpRequestHandler {
+            runtime,
+            invocation_channel,
+        };
+        let server = self.make_server(generic_handler, notify_receiver);
+
+        // Create an Async runtime for executing futures.
+        // https://docs.rs/tokio/
+        let mut async_runtime = create_async_runtime();
+
+        // Start the HTTP server.
+        info!(
+            "{}: Starting HTTP server pseudo-Node on: {}",
+            self.node_name, self.address
+        );
+        async_runtime.block_on(server);
+    }
+}
+
+fn create_async_runtime() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new()
+        // Use simple scheduler that runs all tasks on the current-thread.
+        // https://docs.rs/tokio/0.2.16/tokio/runtime/index.html#basic-scheduler
+        .basic_scheduler()
+        // Enables the I/O driver.
+        // Necessary for using net, process, signal, and I/O types on the Tokio runtime.
+        .enable_io()
+        // Enables the time driver.
+        // Necessary for creating a Tokio Runtime.
+        .enable_time()
+        .build()
+        .expect("Couldn't create Async runtime")
+}
+
+/// [`HttpRequestHandler`] handles HTTP requests from a client and sends HTTP responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to the Runtime in the context of this HTTP server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing HTTP invocations.
+    invocation_channel: oak_abi::Handle,
+}
+
+impl HttpRequestHandler {
+    fn handle(
+        &self,
+        request: HttpRequest,
+    ) -> Pin<Box<dyn Future<Output = Result<Response<Body>, OakStatus>> + Send + Sync>> {
+        let handler = self.clone();
+        let future = async move {
+            let oak_label = get_oak_label(&request)?;
+            info!(
+                "handling HTTP/2 request; request size: {} bytes, label: {:?}",
+                request.body.len(),
+                oak_label
+            );
+
+            debug!("inject the request into the Oak Node");
+            let response = handler
+                .inject_http_request(request, &oak_label)
+                .map_err(|_| OakStatus::ErrInternal)?;
+
+            Ok(response.to_response())
+        };
+
+        Box::pin(future)
+    }
+
+    fn inject_http_request(
+        &self,
+        request: HttpRequest,
+        label: &Label,
+    ) -> Result<HttpResponseIterator, ()> {
+        // Create a pair of temporary channels to pass the HTTP/2 request and to receive the
+        // response.
+        let pipe = Pipe::new(&self.runtime.clone(), label)?;
+
+        // Serialize HTTP/2 request into a message.
+        let mut message = crate::NodeMessage {
+            data: vec![],
+            handles: vec![],
+        };
+
+        request.encode(&mut message.data).map_err(|err| {
+            error!("Couldn't serialize HttpRequest message: {}", err);
+        })?;
+
```

Comment:
I thought that protobuf objects already implement encode / decode, or is that only for the SDK types? @conradgrobler I think you are looking into unifying those anyways, is that correct?

Reply:
Encodable and Decodable is already implemented in the runtime for Prost Messages. It is different from the version is in the SDK, but it is implemented.

I have looked into unifying the two, but backed out of the change for now as it required moving more than half of the code out of the SDK into the ABI due to related dependencies. There are also some dependencies that are problematic to merge.